packages/node-core/src/indexer/dictionary/dictionary.service.ts

```diff
@@ -0,0 +1,126 @@
+    protected readonly eventEmitter: EventEmitter2
+  ) {}
+
+  abstract initDictionaries(): Promise<void>;
+
+  init(dictionaries: D[]): void {
+    this._dictionaries = dictionaries;
+  }
+
+  private getDictionary(height: number): D | undefined {
+    if (this._dictionaries.length === 0) {
+      logger.debug(`No dictionaries available to use`);
+      return undefined;
+    }
+    // If current dictionary is valid, use current one instead of find a dictionary
+    if (
+      this._currentDictionaryIndex !== undefined &&
+      this._dictionaries[this._currentDictionaryIndex].heightValidation(height)
+    ) {
+      return this._dictionaries[this._currentDictionaryIndex];
+    } else {
+      this.findDictionary(height);
+      if (this._currentDictionaryIndex === undefined) {
+        logger.warn(`No supported dictionary found`);
+        return undefined;
+      } else {
+        logger.debug(`Updated : current dictionary Index is ${this._currentDictionaryIndex}`);
+        return this._dictionaries[this._currentDictionaryIndex];
+      }
+    }
+  }
+
+  // Find the next valid dictionary
+  private findDictionary(height: number, skipDictionaryIndex: number[] = []) {
+    // remove dictionary not valid
+    this._dictionaries = this._dictionaries.filter((d) => d.heightValidation(height));
+    this._currentDictionaryIndex = this._dictionaries.findIndex((d) => d.metadataValid);
+  }
+
+  useDictionary(height: number): boolean {
+    return !!this.getDictionary(height);
+  }
+
+  /**
+   *
+   * @param dataSources
+   */
+
+  buildDictionaryEntryMap(dataSources: BlockHeightMap<DS[]>): void {
+    for (const dict of this._dictionaries) {
+      dict.updateQueriesMap(dataSources);
+    }
+  }
+
+  async scopedDictionaryEntries(
+    startBlockHeight: number,
+    scaledBatchSize: number,
+    latestFinalizedHeight: number //api FinalizedHeight
+  ): Promise<(DictionaryResponse<number | IBlock<FB>> & {queryEndBlock: number}) | undefined> {
+    const dictionary = this.getDictionary(startBlockHeight);
+    if (!dictionary) return undefined;
+
+    /* queryEndBlock needs to be limited by the latest height or the maximum value of endBlock in datasources.
+     * Dictionaries could be in the future depending on if they index unfinalized blocks or the node is using an RPC endpoint that is behind.
+     *
+     * DictionaryV1 should use start + dictionaryQuerySize compare with api latestFinalizedHeight
+     * DictionaryV2 should use start + dictionaryQuerySize compare with dictionary metadata endHeight( same as lastProcessedHeight)
+     *
+     * */
+    const queryEndBlock = dictionary.getQueryEndBlock(startBlockHeight, latestFinalizedHeight);
+
+    const dict = await dictionary.getData(startBlockHeight, queryEndBlock, scaledBatchSize);
+    // Check undefined
+    if (!dict) return undefined;
+
+    // Return the queryEndBlock to know if the scoped entry changed it.
+    return {
+      ...dict,
+      queryEndBlock,
```

Comment:
I think this may be conflicting with https://github.com/subquery/subql-ethereum/blob/2d5dd39ef5c2c9fbb18582213e05d06d5f1355f7/packages/node/src/indexer/dictionary/v2/ethDictionaryV2.ts#L292

In this case here, I think dictionary v1 should set the queryEndBlock or queryEndBlock needs to be merged with lastBufferedHeight